File: src/main/java/edu/berkeley/veloxms/storage/TachyonStorage.java
package edu.berkeley.veloxms.storage;


import org.apache.commons.lang3.SerializationUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import tachyon.r.sorted.ClientStore;

import java.io.IOException;
import java.util.HashMap;
import java.nio.ByteBuffer;

public class TachyonStorage implements ModelStorage {

    private final ClientStore users;
    private final ClientStore items;
    private final ClientStore ratings;
    private final ClientStore matPredictions;
    private final int numFactors;

    private static final Logger LOGGER = LoggerFactory.getLogger(TachyonStorage.class);

    // TODO eventually we will want a shared cache among resources
    /* private final ConcurrentHashMap<Long, double[]> itemCache; */
    /* private final ConcurrentHashMap<Long, double[]> userCache; */

    public TachyonStorage(ClientStore users,
                          ClientStore items,
                          ClientStore ratings,
                          ClientStore matPredictions,
                          int numFactors) {
        this.users = users;
        this.items = items;
        this.ratings = ratings;
        this.matPredictions = matPredictions;
        this.numFactors = numFactors;
    }

    @Override
    public double[] getItemFactors(long itemId) {
        return getFactors(itemId, items);
    }

    @Override
    public double[] getUserFactors(long userId) {
        return getFactors(userId, users);
    }

    private static double[] getFactors(long id, ClientStore model) {
        // ByteBuffer key = ByteBuffer.allocate(8); 
        // key.putLong(id); 
        try {
            byte[] rawBytes = model.get(TachyonUtils.long2ByteArr(id));
            return (double[]) SerializationUtils.deserialize(rawBytes);
        } catch (IOException e) {
            LOGGER.warn("Caught tachyon exception: " + e.getMessage());
        }
        return null;
    }
    
    @Override
    public HashMap<Long, Float> getRatedMovies(long userId) {
        HashMap<Long, Float> ratedMovies = null;
        try {
            LOGGER.info("Looking for ratings for user: " + userId);
            byte[] rawBytes = ratings.get(TachyonUtils.long2ByteArr(userId));
            if (rawBytes != null) {
                ratedMovies = (HashMap<Long, Float>) SerializationUtils.deserialize(rawBytes);
            } else {
                LOGGER.warn("no value found in ratings for user: " + userId);
            }
        } catch (IOException e) {
            LOGGER.warn("Caught tachyon exception: " + e.getMessage());
        }
        return ratedMovies;
    }

    @Override
    public int getNumFactors() {
        return this.numFactors;
    }


    // TODO deicde if a single KV pair per prediction is the best way to do this
    @Override
    public double getMaterializedPrediction(long userId, long movieId) {
        double prediction = -1.0;
        try {
            byte[] rawPrediction =
                matPredictions.get(TachyonUtils.twoDimensionKey(userId, movieId));
            if (rawPrediction != null) {
                prediction = ByteBuffer.wrap(rawPrediction).getDouble();
            }
        } catch (IOException e) {
            LOGGER.warn("Caught tachyon exception: " + e.getMessage());
        }
        return prediction;
    }
}

File: veloxms-core/src/main/scala/edu/berkeley/veloxms/storage/TachyonStorage.scala
package edu.berkeley.veloxms.storage


import org.apache.commons.lang3.NotImplementedException
import edu.berkeley.veloxms.util.Logging
import tachyon.r.sorted.ClientStore
import tachyon.TachyonURI
import scala.util._
import java.io.IOException
import java.nio.ByteBuffer
import scala.collection.immutable.HashMap
import edu.berkeley.veloxms._
import java.nio.ByteBuffer
import java.io.ByteArrayOutputStream
import java.util.concurrent.ConcurrentHashMap
import edu.berkeley.veloxms.util.{VeloxKryoRegistrar, KryoThreadLocal}

// class TachyonStorage[U: ClassTag] (
class TachyonStorage[K, V] (
                         tachyonMaster: String,
                         storeName: String) extends ModelStorage[K, V] with Logging {

  val store = TachyonUtils.getStore(tachyonMaster, storeName) match {
    case Success(s) => s
    case Failure(f) => throw new RuntimeException(
      s"Couldn't initialize use model: ${f.getMessage}")
  }

  val storeCache = new ConcurrentHashMap[K, V]

  logInfo("got tachyon stores")

  /**
   * Cleans up any necessary resources
   */
  override def stop() {}

  override def put(kv: (K, V)): Unit = storeCache.put(kv._1, kv._2)

  override def get(key: K): Option[V] = {
    val result = Option(storeCache.get(key)).getOrElse({
      val rawBytes = ByteBuffer.wrap(store.get(StorageUtils.toByteArr(key)))
      val kryo = KryoThreadLocal.kryoTL.get
      kryo.deserialize(rawBytes).asInstanceOf[V]
    })
    Some(result)
  }

  // def getAllObservations(userId: Long): Try[Map[Long, Double]] = {
  //
  //   val rawBytes = ByteBuffer.wrap(ratings.get(StorageUtils.long2ByteArr(userId)))
  //   val kryo = KryoThreadLocal.kryoTL.get
  //   val result = kryo.deserialize(rawBytes).asInstanceOf[HashMap[Long, Double]]
  //
  //   Success(TachyonUtils.mergeObservations(Option(result), Option(observationsCache.get(userId))))
  //     //
  //     // val result = for {
  //     //     rawBytes <- Try(ratings.get(TachyonUtils.long2ByteArr(userId)))
  //     //     array <- Try(rawBytes.unpickle[Map[Long, Double]])
  //     // } yield array
  //     // result
  // }
}



object TachyonUtils {

  def getStore(tachyon: String, kvloc: String): Try[ClientStore] = {
    val url = s"$tachyon/$kvloc"
    Try(ClientStore.getStore(new TachyonURI(url))) match {
      case f: Failure[ClientStore] => f
      case Success(u) => if (u == null) {
        Failure(new RuntimeException(s"Tachyon store $url not found"))
      } else {
        Success(u)
      }
    }
  }

  def mergeObservations[T]( tachyonMap: Option[HashMap[T, Double]],
                         cacheEntry: Option[HashMap[T, Double]]): HashMap[T, Double] = {
    if (!cacheEntry.isDefined) {
      tachyonMap.getOrElse({
        new HashMap[T, Double]
      })
    } else if (!tachyonMap.isDefined) {
      cacheEntry.get
    } else {
      var result = tachyonMap.get

      for (iid <- cacheEntry.get.keys)  {
        if (result.keySet.contains(iid)) { // if there is an entry, then remove it
          result = result - iid
        }

        val cacheEntryVal: Double = cacheEntry.get.get(iid) match {
          case Some(cacheEntryVal) => cacheEntryVal
          case None => throw new RuntimeException("iid in key set not found on get")
        }

        result = result + (iid -> cacheEntryVal)
      }

      result
    }
  }

}


